// @keqi.gress/plugin-dev-server/config/gress-dev.config.example.js
/**
 * Gress 插件开发服务器配置示例
 * 
 * 将此文件复制到插件前端根目录，重命名为 gress-dev.config.js
 */

export default {
  // ==================== 基础配置 ====================
  
  /** 插件名称（必填） */
  pluginName: 'appstore',
  
  /** 后端服务地址（可选，默认 http://localhost:8080） */
  backendUrl: 'http://localhost:8080',
  
  /** 开发服务器端口（可选，默认 3000） */
  port: 3001,
  
  /** API 前缀（可选，默认 /api/plugin/{pluginName}） */
  apiPrefix: '/api/plugin/appstore',
  
  // ==================== 路由配置 ====================
  
  /**
   * 是否自动扫描 views 目录生成路由（可选，默认 true）
   * 
   * 开启后会自动扫描 src/views 目录下的所有 .vue 文件，
   * 根据文件名和目录结构自动生成路由配置。
   * 
   * 规则：
   * - views/HomePage.vue → path: '/'
   * - views/UserList.vue → path: '/user-list'
   * - views/user/Detail.vue → path: '/user/detail'
   * - views/[id].vue → path: '/:id'
   */
  autoScanRoutes: true,
  
  /**
   * 手动配置的路由（可选）
   * 
   * 这里配置的路由会覆盖自动扫描的路由。
   * 适用于需要自定义路径、元信息的场景。
   */
  devRoutes: [
    // 示例：自定义首页路由
    // {
    //   path: '/',
    //   component: 'Dashboard',  // 使用 Dashboard 而不是 HomePage
    //   meta: {
    //     title: '控制台',
    //     icon: 'Dashboard',
    //     keepAlive: true
    //   }
    // },
    
    // 示例：自定义路径
    // {
    //   path: '/applications',  // 覆盖自动生成的 /application-list
    //   component: 'ApplicationList',
    //   meta: {
    //     title: '应用列表',
    //     icon: 'Apps'
    //   }
    // },
    
    // 示例：动态路由
    // {
    //   path: '/applications/:id',
    //   component: 'ApplicationDetail',
    //   meta: {
    //     title: '应用详情'
    //   }
    // },
    
    // 示例：嵌套路由
    // {
    //   path: '/settings',
    //   component: 'Settings',
    //   meta: {
    //     title: '设置'
    //   },
    //   children: [
    //     {
    //       path: 'profile',
    //       component: 'settings/Profile',
    //       meta: { title: '个人资料' }
    //     },
    //     {
    //       path: 'security',
    //       component: 'settings/Security',
    //       meta: { title: '安全设置' }
    //     }
    //   ]
    // }
  ],
  
  /**
   * 菜单配置（可选）
   * 
   * 用于在开发环境模拟后端返回的菜单数据。
   */
  devMenus: [
    // {
    //   id: 'home',
    //   title: '首页',
    //   path: '/',
    //   icon: 'Home'
    // },
    // {
    //   id: 'applications',
    //   title: '应用管理',
    //   path: '/applications',
    //   icon: 'Apps',
    //   children: [
    //     {
    //       id: 'app-list',
    //       title: '应用列表',
    //       path: '/applications',
    //       icon: 'List'
    //     },
    //     {
    //       id: 'app-create',
    //       title: '创建应用',
    //       path: '/applications/create',
    //       icon: 'Add'
    //     }
    //   ]
    // },
    // {
    //   id: 'settings',
    //   title: '设置',
    //   path: '/settings',
    //   icon: 'Settings'
    // }
  ],
  
  // ==================== 服务器配置 ====================
  
  /** 是否启用热更新（可选，默认 true） */
  hmr: true,
  
  /** 是否自动打开浏览器（可选，默认 false） */
  open: false,
  
  /** 路由基础路径（可选，默认 '/'） */
  base: '/',
  
  // ==================== 高级配置 ====================
  
  /**
   * 额外的 Vite 配置（可选）
   * 
   * 可以在这里添加自定义的 Vite 配置，
   * 会与默认配置合并。
   */
  vite: {
    // 示例：配置路径别名
    // resolve: {
    //   alias: {
    //     '@': '/src',
    //     '@components': '/src/components',
    //     '@views': '/src/views'
    //   }
    // },
    
    // 示例：配置 CSS 预处理器
    // css: {
    //   preprocessorOptions: {
    //     scss: {
    //       additionalData: `@import "@/styles/variables.scss";`
    //     }
    //   }
    // },
    
    // 示例：配置构建选项
    // build: {
    //   sourcemap: true,
    //   minify: false
    // }
  },
  
  /**
   * 额外的代理配置（可选）
   * 
   * 除了默认的 API 代理，还可以添加其他代理规则。
   */
  proxy: {
    // 示例：代理其他 API
    // '/api/other': {
    //   target: 'http://localhost:9000',
    //   changeOrigin: true,
    //   rewrite: (path) => path.replace(/^\/api\/other/, '')
    // },
    
    // 示例：代理 WebSocket
    // '/ws': {
    //   target: 'ws://localhost:8080',
    //   ws: true
    // }
  }
}
