// @keqi.gress/plugin-ui/dist/index.js
import { ref, computed, onMounted, onUnmounted, defineComponent, openBlock, createElementBlock, createElementVNode, toDisplayString, createCommentVNode, renderSlot } from "vue";
function useLoadingState(options = {}) {
  const { initialLoading = false, onError } = options;
  const loading = ref(initialLoading);
  const error = ref(null);
  const isEmpty = ref(false);
  const isLoading = computed(() => loading.value);
  const hasError = computed(() => error.value !== null);
  const isEmptyState = computed(() => isEmpty.value && !loading.value && !error.value);
  const startLoading = () => {
    loading.value = true;
    error.value = null;
    isEmpty.value = false;
  };
  const stopLoading = () => {
    loading.value = false;
  };
  const setError = (err) => {
    loading.value = false;
    error.value = err instanceof Error ? err : new Error(err);
    if (onError && error.value) {
      onError(error.value);
    }
  };
  const clearError = () => {
    error.value = null;
  };
  const setEmpty = (empty) => {
    isEmpty.value = empty;
  };
  const reset = () => {
    loading.value = false;
    error.value = null;
    isEmpty.value = false;
  };
  const execute = async (fn, options2) => {
    try {
      startLoading();
      const result = await fn();
      if (options2 == null ? void 0 : options2.checkEmpty) {
        isEmpty.value = options2.checkEmpty(result);
      }
      if (options2 == null ? void 0 : options2.onSuccess) {
        options2.onSuccess(result);
      }
      return result;
    } catch (err) {
      const errorObj = err instanceof Error ? err : new Error(String(err));
      setError(errorObj);
      if (options2 == null ? void 0 : options2.onError) {
        options2.onError(errorObj);
      }
      return null;
    } finally {
      stopLoading();
    }
  };
  return {
    // 状态
    loading,
    error,
    isEmpty,
    // 计算属性
    isLoading,
    hasError,
    isEmptyState,
    // 方法
    startLoading,
    stopLoading,
    setError,
    clearError,
    setEmpty,
    reset,
    execute
  };
}
const breakpoints = {
  xs: 480,
  sm: 640,
  md: 768,
  lg: 1024,
  xl: 1280,
  "2xl": 1536
};
function useResponsive() {
  const windowWidth = ref(typeof window !== "undefined" ? window.innerWidth : 1024);
  const updateWidth = () => {
    windowWidth.value = window.innerWidth;
  };
  onMounted(() => {
    window.addEventListener("resize", updateWidth);
  });
  onUnmounted(() => {
    window.removeEventListener("resize", updateWidth);
  });
  const currentBreakpoint = computed(() => {
    const width = windowWidth.value;
    if (width < breakpoints.xs) return "xs";
    if (width < breakpoints.sm) return "xs";
    if (width < breakpoints.md) return "sm";
    if (width < breakpoints.lg) return "md";
    if (width < breakpoints.xl) return "lg";
    if (width < breakpoints["2xl"]) return "xl";
    return "2xl";
  });
  const isMobile = computed(() => windowWidth.value < breakpoints.md);
  const isTablet = computed(
    () => windowWidth.value >= breakpoints.md && windowWidth.value < breakpoints.lg
  );
  const isDesktop = computed(() => windowWidth.value >= breakpoints.lg);
  const isSmaller = (breakpoint) => {
    return computed(() => windowWidth.value < breakpoints[breakpoint]);
  };
  const isLarger = (breakpoint) => {
    return computed(() => windowWidth.value >= breakpoints[breakpoint]);
  };
  const isBetween = (min, max) => {
    return computed(
      () => windowWidth.value >= breakpoints[min] && windowWidth.value < breakpoints[max]
    );
  };
  const responsive = (values, defaultValue) => {
    const bp = currentBreakpoint.value;
    const breakpointOrder = ["2xl", "xl", "lg", "md", "sm", "xs"];
    const currentIndex = breakpointOrder.indexOf(bp);
    for (let i = currentIndex; i < breakpointOrder.length; i++) {
      const key = breakpointOrder[i];
      if (values[key] !== void 0) {
        return values[key];
      }
    }
    return defaultValue;
  };
  const getResponsiveColumns = (config) => {
    const defaultConfig = {
      xs: 1,
      sm: 1,
      md: 2,
      lg: 3,
      xl: 4,
      "2xl": 4
    };
    const finalConfig = { ...defaultConfig, ...config };
    return responsive(finalConfig, 1);
  };
  const getResponsiveSpacing = (config) => {
    const defaultConfig = {
      xs: 8,
      sm: 12,
      md: 16,
      lg: 24
    };
    const finalConfig = { ...defaultConfig, ...config };
    return responsive(finalConfig, 16);
  };
  return {
    // 状态
    windowWidth,
    currentBreakpoint,
    // 计算属性
    isMobile,
    isTablet,
    isDesktop,
    // 方法
    isSmaller,
    isLarger,
    isBetween,
    responsive,
    getResponsiveColumns,
    getResponsiveSpacing,
    // 断点值
    breakpoints
  };
}
const _hoisted_1$2 = { class: "loading-container" };
const _hoisted_2$2 = {
  key: 0,
  class: "loading-message"
};
const _sfc_main$2 = /* @__PURE__ */ defineComponent({
  __name: "LoadingState",
  props: {
    message: {}
  },
  setup(__props) {
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", _hoisted_1$2, [
        _cache[0] || (_cache[0] = createElementVNode("div", { class: "loading-spinner" }, null, -1)),
        __props.message ? (openBlock(), createElementBlock("p", _hoisted_2$2, toDisplayString(__props.message), 1)) : createCommentVNode("", true)
      ]);
    };
  }
});
const _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};
const LoadingState = /* @__PURE__ */ _export_sfc(_sfc_main$2, [["__scopeId", "data-v-88be2f1a"]]);
const _hoisted_1$1 = { class: "empty-state" };
const _hoisted_2$1 = { class: "empty-icon" };
const _hoisted_3$1 = { class: "empty-title" };
const _hoisted_4$1 = {
  key: 0,
  class: "empty-description"
};
const _hoisted_5$1 = {
  key: 1,
  class: "empty-action"
};
const _sfc_main$1 = /* @__PURE__ */ defineComponent({
  __name: "EmptyState",
  props: {
    title: { default: "暂无数据" },
    description: {}
  },
  setup(__props) {
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", _hoisted_1$1, [
        createElementVNode("div", _hoisted_2$1, [
          renderSlot(_ctx.$slots, "icon", {}, () => [
            _cache[0] || (_cache[0] = createElementVNode("svg", {
              width: "64",
              height: "64",
              viewBox: "0 0 64 64",
              fill: "none",
              xmlns: "http://www.w3.org/2000/svg"
            }, [
              createElementVNode("path", {
                d: "M32 8C18.745 8 8 18.745 8 32s10.745 24 24 24 24-10.745 24-24S45.255 8 32 8zm0 44c-11.028 0-20-8.972-20-20s8.972-20 20-20 20 8.972 20 20-8.972 20-20 20z",
                fill: "currentColor",
                opacity: "0.3"
              }),
              createElementVNode("path", {
                d: "M32 24c-1.105 0-2 .895-2 2v12c0 1.105.895 2 2 2s2-.895 2-2V26c0-1.105-.895-2-2-2zm0 20c-1.105 0-2 .895-2 2s.895 2 2 2 2-.895 2-2-.895-2-2-2z",
                fill: "currentColor"
              })
            ], -1))
          ], true)
        ]),
        createElementVNode("h3", _hoisted_3$1, toDisplayString(__props.title), 1),
        __props.description ? (openBlock(), createElementBlock("p", _hoisted_4$1, toDisplayString(__props.description), 1)) : createCommentVNode("", true),
        _ctx.$slots.action ? (openBlock(), createElementBlock("div", _hoisted_5$1, [
          renderSlot(_ctx.$slots, "action", {}, void 0, true)
        ])) : createCommentVNode("", true)
      ]);
    };
  }
});
const EmptyState = /* @__PURE__ */ _export_sfc(_sfc_main$1, [["__scopeId", "data-v-12ee016f"]]);
const _hoisted_1 = { class: "error-container" };
const _hoisted_2 = { class: "error-icon" };
const _hoisted_3 = { class: "error-title" };
const _hoisted_4 = { class: "error-message" };
const _hoisted_5 = {
  key: 0,
  class: "error-action"
};
const _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "ErrorState",
  props: {
    error: {},
    title: { default: "出错了" },
    showRetry: { type: Boolean, default: false }
  },
  emits: ["retry"],
  setup(__props, { emit: __emit }) {
    const props = __props;
    const emit = __emit;
    const errorMessage = computed(() => {
      if (!props.error) return "发生了未知错误";
      if (typeof props.error === "string") return props.error;
      return props.error.message || "发生了未知错误";
    });
    const handleRetry = () => {
      emit("retry");
    };
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", _hoisted_1, [
        createElementVNode("div", _hoisted_2, [
          renderSlot(_ctx.$slots, "icon", {}, () => [
            _cache[0] || (_cache[0] = createElementVNode("svg", {
              width: "64",
              height: "64",
              viewBox: "0 0 64 64",
              fill: "none",
              xmlns: "http://www.w3.org/2000/svg"
            }, [
              createElementVNode("path", {
                d: "M32 8C18.745 8 8 18.745 8 32s10.745 24 24 24 24-10.745 24-24S45.255 8 32 8zm0 44c-11.028 0-20-8.972-20-20s8.972-20 20-20 20 8.972 20 20-8.972 20-20 20z",
                fill: "currentColor",
                opacity: "0.3"
              }),
              createElementVNode("path", {
                d: "M38.828 22.172a2 2 0 00-2.828 0L32 26.172l-4-4a2 2 0 10-2.828 2.828L29.172 29l-4 4a2 2 0 102.828 2.828L32 31.828l4 4a2 2 0 102.828-2.828L34.828 29l4-4a2 2 0 000-2.828z",
                fill: "currentColor"
              })
            ], -1))
          ], true)
        ]),
        createElementVNode("h3", _hoisted_3, toDisplayString(__props.title), 1),
        createElementVNode("p", _hoisted_4, toDisplayString(errorMessage.value), 1),
        _ctx.$slots.action || __props.showRetry ? (openBlock(), createElementBlock("div", _hoisted_5, [
          renderSlot(_ctx.$slots, "action", {}, () => [
            __props.showRetry ? (openBlock(), createElementBlock("button", {
              key: 0,
              onClick: handleRetry,
              class: "error-retry-btn"
            }, " 重试 ")) : createCommentVNode("", true)
          ], true)
        ])) : createCommentVNode("", true)
      ]);
    };
  }
});
const ErrorState = /* @__PURE__ */ _export_sfc(_sfc_main, [["__scopeId", "data-v-9bbae138"]]);
export {
  EmptyState,
  ErrorState,
  LoadingState as LoadingStateComponent,
  breakpoints,
  useLoadingState,
  useResponsive
};
